### PR TITLE
Use the latest CF7 JS events

### LINKED ISSUE #10: Update JS form event triggers since on_sent_ok/on_submit have been deprecated
https://contactform7.com/additional-settings/#javascript shows that the developer of Contact Form 7 is planning to drop support of on_sent_ok/OnSentOk and on_submit/OnSubmit by EOY 2017.

The successful page redirect and Google Analytics event trigger features both rely on the use of these deprecated commands and will therefore break as soon as the version which removes this is released. I'm not sure if any other features use this, but those are two I know of.

It would be best to rewrite these features so they no longer use on_sent_ok/OnSentOk or on_submit/OnSubmit sooner than later.

I have also posted this on the WordPress.org support forum at: https://wordpress.org/support/topic/redirection-analytics-event-features-rely-on-deprecated-cf7-functionality/

### PR DESCRIPTION
**Pull request by kasparsd**
- Fixes #10
- Replaces #12

### CODE REVIEW

File: plugin.php [hunk 1]
```diff
@@ -556,6 +545,8 @@ function wpcf7_submit( $form, $result ) {
 		// JS is already doing the redirect
 		if ( isset( $_POST['_wpcf7_is_ajax_call'] ) || ! isset( $result['status'] ) ) {
 			return;
+		} elseif ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
```
Comment (kasparsd) [inline plugin.php]: `_wpcf7_is_ajax_call` was removed in the version of CF7 that switched to using REST calls for submissions instead.

File: plugin.php [hunk 2]
```diff
@@ -496,57 +485,57 @@ function dequeue_styles() {
...
-			$items['onSubmit'][] = sprintf(
-					'if ( typeof ga == "function" ) {
-						ga( "send", "event", "Contact Form", "Submit", "%1$s" );
-					}
-					if ( typeof _gaq !== "undefined" ) {
-						_gaq.push([ "_trackEvent", "Contact Form", "Submit", "%1$s" ]);
-					}',
-					esc_js( $form->title() )
-				);
+	function track_form_events() {
```
Comment (kasparsd) [inline plugin.php]: Collect a list of forms that will trigger analytics events and/or redirects and make them available to our new JS.

File: plugin.php [hunk 3]
```diff
@@ -420,13 +403,19 @@ function get_form_settings( $form, $field = null, $fresh = false ) {
...
+				'track-ga' => false,
 				'google-recaptcha-lang' => null,
 			)
 		);
 
 		// Cache it for re-use
 		$form_settings[ $form->id() ] = $settings;
 
+		// Convert individual legacy settings into one.
+		if ( ! empty( $settings['track-ga-success'] ) || ! empty( $settings['track-ga-submit'] ) ) {
```
Comment (kasparsd) [inline plugin.php]: Combine GA-related settings into one for simplicity.

File: js/controls.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		formId = parseInt( formId );
+
+		if ( window.cf7_extras.forms && window.cf7_extras.forms[ formId ] ) {
+			return window.cf7_extras.forms[ formId ];
+		}
+
+		return false;
+	}
+
+	$( document ).on( 'wpcf7:mailsent', function( event, form ) {
```
Comment (kasparsd) [inline js/controls.js]: Attach listeners to various CF7 events.

**Comment by buzztone** [comment]
It looks like these changes mean that contact-form-7-extras now longer work with some previous versions of Contact Form 7.

The required Contact Form 7 version could be added to readme.txt to help users.
e.g. Requires Contact Form 7 Version: 4.3+

**Comment by kasparsd** [comment]
> The required Contact Form 7 version could be added to readme.txt to help users.

@buzztone That is a great idea! Could you please open an issue for that?

**Pull request merged**